### PR TITLE
table: optimize add record used in lightning

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve?
Optimize table.AddRecord so lightning can run faster. These change can improve lightning performance by about 10% when running with local backend.

Problem Summary:

What's Changed:
- ~~add a new interface `table. AddRecordWithCtx` to reuse the temp slices and bufStore. this can help improve about 5% of lightning total performance.~~
- add a new type `CommonAddRecordCtx` as a `sessionctx.Value` to reuse the temp slices and bufStore. this can help improve about 5% of lightning total performance.
- Change some usage of `Datum` by pointer to avoid the memmove cost. Datum is a struct of 70 bytes, so in some case, use pointer will be faster.  

How it Works:

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

### Release note 
- No release note

### CODE REVIEW

**Comment by glorv** [comment]
@tiancaiamao @coocood PTAL

File: table/tables/tables.go [hunk 1]
```diff
@@ -485,8 +486,31 @@ func FindPrimaryIndex(tblInfo *model.TableInfo) *model.IndexInfo {
...
+	}
+}
+
 // AddRecord implements table.Table AddRecord interface.
 func (t *TableCommon) AddRecord(ctx sessionctx.Context, r []types.Datum, opts ...table.AddRecordOption) (recordID kv.Handle, err error) {
+	return t.AddRecordWithCtx(ctx, r, nil, opts...)
+}
+
+// AddRecordWithCtx implements table.Table AddRecordWithCtx interface.
+func (t *TableCommon) AddRecordWithCtx(ctx sessionctx.Context, r []types.Datum, addCtx interface{}, opts ...table.AddRecordOption) (recordID kv.Handle, err error) {
```
Comment (coocood) [inline table/tables/tables.go:513]: I think we can put the `addCtx` into the `ctx`, then we don't need to add a method to the `Table` interface.

File: types/mydecimal.go [hunk 2]
```diff
@@ -1185,7 +1191,13 @@ func (d *MyDecimal) ToBin(precision, frac int) ([]byte, error) {
 	fracSizeFrom := wordsFracFrom*wordSize + dig2bytes[trailingDigitsFrom]
 	originIntSize := intSize
 	originFracSize := fracSize
-	bin := make([]byte, intSize+fracSize)
+	bufLen := len(buf)
+	if bufLen+intSize+fracSize <= cap(buf) {
+		buf = buf[:bufLen+intSize+fracSize]
+	} else {
+		buf = append(buf, bytes.Repeat([]byte{0}, intSize+fracSize)...)
```
Comment (coocood) [inline types/mydecimal.go:1198]:
```suggestion
		buf = append(buf, make([]byte, intSize+fracSize)...)
```

**Comment by coocood** [comment]
LGTM

**Comment by glorv** [comment]
@lysu  @tiancaiamao  PTAL

**Comment by tiancaiamao** [comment]
LGTM
Please resolve conflicts @glorv

**Pull request merged**